text/0000-jsbox.md

```diff
@@ -0,0 +1,341 @@
+
+```rust
+impl<T> JsBox where T: Send + 'static {
+    pub fn borrow<'a: 'b, 'b, C: Context<'a>>(
+        self,
+        cx: &'b C,
+    ) -> &'b T;
+    pub fn borrow_mut<'a: 'b, 'b, C: Context<'a>>(
+        self,
+        cx: &mut 'b C,
+    ) -> &'b mut T;
+}
+```
+
+Borrowing rules on `Context` would enforce that only a single `JsBox` could be borrowed mutably at once. However, this approach has two significant drawbacks that impair ergonomics:
+
+* Many `neon` methods require `&mut Context`. While a `JsBox` is borrowed, most `neon` methods could not be called. We could potentially mitigate this by performing a sweeping audit and relaxing constraints where a shared reference is acceptable
+* Multiple `JsBox`, even if they are unrelated, could not be borrowed at the same time because there is no way to statically prove they do not point to the same underlying data
+* Impairs ergonomics on APIs that do not require mutable borrowing since `Deref` cannot be implemented
+
+### `JsCell`
+
+Neon could provide a `JsCell` type that is identical to `JsBox`, but also wraps the underlying `T` in a `std::cell::RefCell`. This allows the user to also mutably borrow the value.
+
+```rust
+impl<T> JsCell where T: Send + 'static {
+    /// This should only fail if the VM is in a throwing state or is
+    /// in the process of stopping.
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: T,
+    ) -> NeonResult<JsCell>;
+    /// May panic if already mutably borrowed
+    pub fn borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Ref<'a, T>;
+    pub fn try_borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<Ref<'a, T>, BorrowError>;
+    /// May panic if already borrowed
+    pub fn borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> RefMut<'a, T>;
+    pub fn try_borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<RefMut<'a, T>, BorrowMutError>;
+    /// The most common `RefCell` methods are provided on `JsCell` as a
+    /// convenience. Other methods may can be accessed by calling `as_cell`
+    pub fn as_cell<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> &RefCell<T>;
+}
+```
+
+This approach has several significant drawbacks:
+
+* Increases the API substantial
+* Adds an abstraction that not all users might need
+* Prevents implementing `Deref`
+
+Instead, Neon can lean on documentation to guide users towards `RefCell` if they need interior mutability.
+
+### Naming
+
+Several other names were considered:
+
+* `JsExternal`. Matches the name in N-API documentation (external reference); however, `cx.external(v)` might be confusing.
+* `JsRef`. Also, matches N-API but, does not match Rust naming conventions.
+* `JsCell`. Only appropriate for the `RefCell` approach.
```

Comment:
Would it make sense to consider a high-level convenience type, something like `type JsRefCell<T> = JsBox<JsRefCell<T>>;`, for the common case of interior mutability? This would make a common case more concise, e.g. in your `Pool` example above, it would change from
```rust
     let mut pool = cx.argument::<JsBox<RefCell<Pool>>>(0)?;
```
to
```rust
     let mut pool = cx.argument::<JsRefCell<Pool>>(0)?;
```
It would also create an opportunity to document the use case of interior mutability directly in the API docs, making the idiom a little more discoverable.

Maybe not necessary for this RFC, but we could leave a note in "unresolved questions" about future convenience APIs and mention the idea there.

Reply:
I added a section about future extension that might include this. In my opinion, this is above and beyond this RFC since in addition to the type alias we would want other helper methods (e.g., `cx.ref_cell(..)`)